File: examples/java/kora-java-database-jdbc/src/main/java/io/koraframework/example/jdbc/JdbcMapperColumnRepository.java
package io.koraframework.example.jdbc;

import org.jspecify.annotations.Nullable;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;
import io.koraframework.common.annotation.Mapping;
import io.koraframework.database.common.annotation.Column;
import io.koraframework.database.common.annotation.Query;
import io.koraframework.database.common.annotation.Repository;
import io.koraframework.database.jdbc.annotation.EntityJdbc;
import io.koraframework.database.jdbc.JdbcRepository;
import io.koraframework.database.jdbc.mapper.parameter.JdbcParameterColumnMapper;
import io.koraframework.database.jdbc.mapper.result.JdbcResultColumnMapper;

@Repository
public interface JdbcMapperColumnRepository extends JdbcRepository {

    final class EntityFieldTypeResultMapper implements JdbcResultColumnMapper<Entity.FieldType> {

        private static final Entity.FieldType[] ALL = Entity.FieldType.values();

        @Override
        public Entity.FieldType apply(ResultSet rs, int index) throws SQLException {
            var fieldAsInt = rs.getInt(index);
            for (var type : ALL) {
                if (type.code() == fieldAsInt) {
                    return type;
                }
            }

            return Entity.FieldType.UNKNOWN;
        }
    }

    final class EntityFieldTypeParameterMapper implements JdbcParameterColumnMapper<Entity.FieldType> {

        @Override
        public void set(PreparedStatement stmt, int index, Entity.@Nullable FieldType value) throws SQLException {
            if (value != null) {
                stmt.setInt(index, value.code());
            }
        }
    }

    @EntityJdbc
    record Entity(String id,
                  @Mapping(EntityFieldTypeResultMapper.class)
                  @Mapping(EntityFieldTypeParameterMapper.class)
                  @Column("value1") FieldType field1,
                  String value2,
                  @Nullable String value3) {

        public enum FieldType {

            UNKNOWN(-10),
            ONE(1),
            TWO(2);

            private final int code;

            FieldType(int code) {
                this.code = code;
            }

            public int code() {
                return code;
            }
        }
    }

    @Query("SELECT * FROM entities")
    List<Entity> findAll();
}

File: examples/kotlin/kora-kotlin-database-jdbc/src/main/kotlin/io/koraframework/kotlin/example/jdbc/JdbcMapperColumnRepository.kt
package io.koraframework.kotlin.example.jdbc

import org.postgresql.util.PGobject
import io.koraframework.common.annotation.Component
import io.koraframework.common.annotation.Mapping
import io.koraframework.common.annotation.Module
import io.koraframework.common.annotation.Root
import io.koraframework.database.common.UpdateCount
import io.koraframework.database.common.annotation.*
import io.koraframework.database.jdbc.annotation.EntityJdbc
import io.koraframework.database.jdbc.JdbcRepository
import io.koraframework.database.jdbc.mapper.parameter.JdbcParameterColumnMapper
import io.koraframework.database.jdbc.mapper.result.JdbcResultColumnMapper
import io.koraframework.database.jdbc.mapper.result.JdbcResultSetMapper
import io.koraframework.database.jdbc.mapper.result.JdbcRowMapper
import io.koraframework.json.common.JsonReader
import io.koraframework.json.common.JsonWriter
import io.koraframework.json.common.annotation.Json
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.Types

@Repository
interface JdbcMapperColumnRepository : JdbcRepository {

    @EntityJdbc
    data class Entity(
        val id: String,
        @Mapping(JdbcEntityFieldTypeResultMapper::class)
        @Mapping(JdbcEntityFieldTypeColumnMapper::class)
        @field:Column("value1")
        val field1: FieldType,
        val value2: String,
        val value3: String?
    ) {
        enum class FieldType(val code: Int) {
            UNKNOWN(-10),
            ONE(1),
            TWO(2)
        }
    }

    @Query("SELECT * FROM entities")
    fun findAll(): List<Entity>
}

class JdbcEntityFieldTypeResultMapper : JdbcResultColumnMapper<JdbcMapperColumnRepository.Entity.FieldType> {
    override fun apply(rs: ResultSet, index: Int): JdbcMapperColumnRepository.Entity.FieldType {
        val fieldAsInt = rs.getInt(index)
        return JdbcMapperColumnRepository.Entity.FieldType.entries.firstOrNull { it.code == fieldAsInt }
            ?: JdbcMapperColumnRepository.Entity.FieldType.UNKNOWN
    }
}

class JdbcEntityFieldTypeColumnMapper : JdbcParameterColumnMapper<JdbcMapperColumnRepository.Entity.FieldType> {
    override fun set(stmt: PreparedStatement, index: Int, value: JdbcMapperColumnRepository.Entity.FieldType?) {
        if (value == null) {
            stmt.setNull(index, Types.INTEGER)
        } else {
            stmt.setInt(index, value.code)
        }
    }
}

